docs/developers_guide/FlinkComponents.md

```diff
@@ -1,22 +1,86 @@
 # Flink components
- 
-Sources, sinks and custom transformations are based on 
-Flink API. Access to various helpers is provided by 
-`FlinkCustomNodeContext`. Special care should be taken to handle
-- lifecycle
-- exception handling properly
 
-:warning: **Flink components should not extend Lifecycle** - it won't be handled properly
+[Sources, sinks and custom transformations](./Basics.md#components-and-componentproviders) are based on Flink API.
+In order to implement any of those you need to provide:
+- a Flink function or Flink `DataStream` transformation
+- a Nussknacker [specification](./Components.md#specification)
 
 ## Sources
 
-Source are defined with [FlinkSource](https://github.com/TouK/nussknacker/blob/staging/engine/flink/components-api/src/main/scala/pl/touk/nussknacker/engine/flink/api/process/FlinkSource.scala).
-In most cases (when you only pass one variable to initial `Context`) it's easier to use `BaseFlinkSource`.
+Implementing a fully functional source ([BasicFlinkSource](https://github.com/TouK/nussknacker/blob/staging/engine/flink/components-api/src/main/scala/pl/touk/nussknacker/engine/flink/api/process/FlinkSource.scala))
+is more complicated since the following things has to be provided:
+- a Flink [SourceFunction](https://nightlies.apache.org/flink/flink-docs-stable/api/java/org/apache/flink/streaming/api/functions/source/SourceFunction.html)
+- Flink [type information](https://nightlies.apache.org/flink/flink-docs-stable/docs/dev/datastream/fault-tolerance/serialization/types_serialization/#flinks-typeinformation-class)
+  for serializing/deserializing the emitted `#input`
+- a [timestamp watermark handler](https://github.com/TouK/nussknacker/blob/staging/engine/flink/components-api/src/main/scala/pl/touk/nussknacker/engine/flink/api/timestampwatermark/TimestampWatermarkHandler.scala)
+  so that events are correctly processed downstream, for example to avoid (or force!) dropping late events by aggregates. Read more about
+  [notion of time](../scenarios_authoring/Intro.md#streaming-flink-only-notion-of-time)
+  and [watermarks](https://nightlies.apache.org/flink/flink-docs-stable/docs/dev/datastream/event-time/generating_watermarks/)
+- (optionally) generating test data support (trait `FlinkSourceTestSupport`) to ease [scenarios authoring](../scenarios_authoring/TestingAndDebugging.md)
+- (optionally) if the source returns a record having generic type and its fields and methods cannot be discovered using Java reflection,
```

Comment:
We want to get rid of `ReturningType` in the end, let's not describe it, leaving only GenericNodeTransformation